src/storage/hummock_trace/src/write.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write;
+use std::mem::size_of;
+
+use bincode::{config, encode_into_std_write};
+#[cfg(test)]
+use mockall::{automock, mock};
+
+use super::record::Record;
+use crate::error::Result;
+use crate::TraceError;
+
+pub(crate) type MagicBytes = u32;
+pub(crate) static MAGIC_BYTES: MagicBytes = 0x484D5452; // HMTR
+
+#[cfg_attr(test, automock)]
+pub(crate) trait TraceWriter {
```

Comment:
Will we have code that takes only the abstract `TraceWriter` instead of the ` TraceWriterImpl`? If not, I think we only need an `TraceWriter` struct instead of a trait and an impl.

Reply:
We may have a more readable serializer and writer.